### ISSUE TITLE
Significant performance drop in pniio 1.3.1 vs 1.1.0

### ISSUE THREAD

**Issue by Ingvord** [pni-libraries/libpniio#156]
Hi,

we are migrating to pninexus (still on pniio though), specifically 1.3.1

Trying the new package on a production machine we have observed a significant performance degradation x1000 times:

| pni stack (core, io) | 1M | 1K |
|--------------------------|------|-----|
|1.1.1, 1.3.1|forever|~33000|
|1.0.0, 1.1.0| ~33000 | 33 |

Here I compared two pni stacks for writing 1M and 1K records. Results are in ms

Tests are mostly identical:

[1.1.1, 1.3.1](https://github.com/hereon-wpi/libpniio-jni/blob/master/jni_cpp/test/main.cpp)

[1.0.0, 1.1.0](https://github.com/hereon-wpi/libpniio-jni/blob/backport/jni_cpp/test/main.cpp)

Basically there is a loop that grows and writes in to a particular field.

Obviously I am doing something wrong while writing with the new libraries stack. Could you please advise?

Thanx in advance

**Comment by Ingvord** [comment]
@feinfinger

FYI

**Comment by jkotan** [comment]
Hi Igor,

thanks for your tests. I will look at it  tomorrow.  Could you please provide your `test.nxdl.xml` file ?

**Comment by yuelongyu** [comment]
Hi Igor,

Compare to pniio 1.1.0, in 1.3.1, nexus::get_objects is called in for loop via get_dataset, basically every write. I suspect this is not very efficient. Can you put it outside the loop and try?

**Comment by Ingvord** [comment]
@jkotan 

Here are test files

For 1.1.0: [test.nxdl.xml.zip](https://github.com/pni-libraries/libpninexus/files/7371730/test.nxdl.xml.zip)

For 1.3.1 : [test.nxdl.xml.zip](https://github.com/pni-libraries/libpninexus/files/7371775/test.nxdl.xml.zip)

@yuelongyu 

Good catch! Here are updated results (I also write only one record now, not two, hence lower numbers):

| ver | 1M (in-loop) | 1M (out-loop)|  1K (in-loop)  | 1K (out-loop) |
|------|------------------|------------------|-------------------|------|
|1.3.1| forever   |  25586      |  17407      |  39-60   |
|1.1.0| forever   |  14979      | 372 | 15   |

Anyway there is x2 times performance drop (out-loop). Unfortunately our use case implies in-loop scenario...

I am thinking about some kind of a map nxpath <-> dataset. Do you know if it is save to keep dataset objects in a map? Won't be there any conflict between opened datasets?

**Comment by jkotan** [comment]
Hi Igor,

Do you use the same libhdf5 in both tests?

**Comment by jkotan** [comment]
Sure. You can keep it in the map (kind of cache). Should be no problem.

**Comment by Ingvord** [comment]
@jkotan 

1.1.0: 

```
        linux-vdso.so.1 (0x00007fff7d6d4000)
        libpniio.so.1.1.0 => /usr/lib/x86_64-linux-gnu/libpniio.so.1.1.0 (0x00007f194ee02000)
        libhdf5_serial.so.100 => /usr/lib/x86_64-linux-gnu/libhdf5_serial.so.100 (0x00007f194e89a000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f194e680000)
        libboost_filesystem.so.1.62.0 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.62.0 (0x00007f194e467000)
        libpnicore.so.1.0.0 => /usr/lib/x86_64-linux-gnu/libpnicore.so.1.0.0 (0x00007f194e1e3000)
        libboost_program_options.so.1.62.0 => /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.62.0 (0x00007f194df64000)
        libboost_regex.so.1.62.0 => /usr/lib/x86_64-linux-gnu/libboost_regex.so.1.62.0 (0x00007f194dc4b000)
        libboost_system.so.1.62.0 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.62.0 (0x00007f194da47000)
        libh5cpp.so.0.4.1 => /usr/lib/x86_64-linux-gnu/libh5cpp.so.0.4.1 (0x00007f194d7b8000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f194d59b000)
        libsz.so.2 => /usr/lib/x86_64-linux-gnu/libsz.so.2 (0x00007f194d398000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f194d194000)
        libhdf5_serial_hl.so.100 => /usr/lib/x86_64-linux-gnu/libhdf5_serial_hl.so.100 (0x00007f194cf70000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f194cc6c000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f194c8ea000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f194c6d3000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f194c334000)
        libboost_date_time.so.1.62.0 => /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.62.0 (0x00007f194c123000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f194bf1b000)
        libicudata.so.57 => /usr/lib/x86_64-linux-gnu/libicudata.so.57 (0x00007f194a49e000)
        libicui18n.so.57 => /usr/lib/x86_64-linux-gnu/libicui18n.so.57 (0x00007f194a023000)
        libicuuc.so.57 => /usr/lib/x86_64-linux-gnu/libicuuc.so.57 (0x00007f1949c7b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f194fe71000)
        libaec.so.0 => /usr/lib/x86_64-linux-gnu/libaec.so.0 (0x00007f1949a73000)
```

1.3.1:

```
        linux-vdso.so.1 (0x00007ffff5389000)
        libpniio.so.1.3.1 => /usr/lib/x86_64-linux-gnu/libpniio.so.1.3.1 (0x00007f1333dda000)
        libhdf5_serial.so.103 => /usr/lib/x86_64-linux-gnu/libhdf5_serial.so.103 (0x00007f1333a55000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f1333837000)
        libboost_filesystem.so.1.67.0 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.67.0 (0x00007f1333819000)
        libpnicore.so.1.1.1 => /usr/lib/x86_64-linux-gnu/libpnicore.so.1.1.1 (0x00007f1333792000)
        libboost_program_options.so.1.67.0 => /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.67.0 (0x00007f133370b000)
        libboost_regex.so.1.67.0 => /usr/lib/x86_64-linux-gnu/libboost_regex.so.1.67.0 (0x00007f13335f4000)
        libboost_system.so.1.67.0 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.67.0 (0x00007f13335ed000)
        libh5cpp.so.0.4.0 => /usr/lib/x86_64-linux-gnu/libh5cpp.so.0.4.0 (0x00007f1333564000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1333543000)
        libsz.so.2 => /usr/lib/x86_64-linux-gnu/libsz.so.2 (0x00007f1333340000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f133333b000)
        libhdf5_serial_hl.so.100 => /usr/lib/x86_64-linux-gnu/libhdf5_serial_hl.so.100 (0x00007f1333313000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1333190000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f133300c000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f1332ff2000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1332e31000)
        libboost_date_time.so.1.67.0 => /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.67.0 (0x00007f1332e1d000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f1332e11000)
        libicudata.so.63 => /usr/lib/x86_64-linux-gnu/libicudata.so.63 (0x00007f1331421000)
        libicui18n.so.63 => /usr/lib/x86_64-linux-gnu/libicui18n.so.63 (0x00007f1331146000)
        libicuuc.so.63 => /usr/lib/x86_64-linux-gnu/libicuuc.so.63 (0x00007f1330f77000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1333f57000)
        libaec.so.0 => /usr/lib/x86_64-linux-gnu/libaec.so.0 (0x00007f1330d6f000)

```

**Comment by jkotan** [comment]
Hi Igor, 
you have a different tests. For 1.3.1 you writer 2 fields 
```c++
    auto dataset1 = get_dataset("/entry/long/value", file);
    auto dataset2 = get_dataset("/entry/long/time", file);
``` 
while for 1.1.0 you writes only one field i.e.
```c++
auto value_fld = get_object(root_group, nxpath::from_string("/entry/long/value"));
```
That is why you have factor 2x.

**Comment by Ingvord** [comment]
Yeah, that was in old version of test, now they are both writing one field. I have committed the new version just now

**Comment by jkotan** [comment]
@Ingvord Indeed, I have to do more tests.

**Comment by jkotan** [comment]
Hi @Ingvord,

I've performed a few benchmarks an I've found that the different in the speed comes from two issue, namely
1) creation of default parameters and additional checks in the h5cpp Dataset write method
2) different default chunking in pniio 1.1.0-1.1.1 (i.e. default chunk =1024) than in pniio 1.3.1-1.3.2. (default chunk =1).

Could you please try your tests by replacing the content of your main loop in the main.cpp (1.3.1) file
```c++
    auto dataset = get_dataset("/entry/long/value", file);
    
    hdf5::dataspace::Hyperslab selection{{0},{1},{1},{1}};

    auto dtype = hdf5::datatype::create<long int>();
    hdf5::property::DatasetTransferList dtpl;
    hdf5::dataspace::Simple simple_space({1}, {hdf5::dataspace::Simple::UNLIMITED});
    
    auto start = chrono::system_clock::now();    
    for(size_t i = 0; i < 1000000; i++) {
      dataset.extent(0, 1);
      auto file_space = dataset.dataspace();
      selection.offset(0, i);
      file_space.selection(hdf5::dataspace::SelectionOperation::SET,selection);
      dataset.write(static_cast<long int>(std::rand()), dtype, simple_space, file_space, dtpl);
    }
    file.flush(hdf5::file::Scope::GLOBAL);
    auto end = chrono::system_clock::now();

    cout << "Time elapsed (s): "
	 << std::chrono::duration<double>(end - start).count()
	 << endl;
```
and the test.nxdl.xml file
```patch
diff --git a/test.nxdl.xml b/test.nxdl.xml
index 6187478..caa46e6 100644
--- a/test.nxdl.xml
+++ b/test.nxdl.xml
@@ -16,7 +16,11 @@
         <field type="uint64" name="time"/>
     </group>
     <group type="NXlog" name="long">
-        <field type="int64" name="value"/>
+        <field type="int64" name="value">
+          <chunk rank="1">
+            <dim value="1024" index="0"/>
+          </chunk>
+        </field>
         <field type="uint64" name="time"/>
     </group>
     <group type="NXlog" name="int">
```
For me it speeds up the writing. The code I send you is not so convenient therefore I will open an h5cpp issue to speed up commands with default parameters.

**Comment by Ingvord** [comment]
@jkotan ,

thank you very much for this code snippet! I have run this in my test environment, here is the result:

|   1M  | 1K |
|--------|------|
| 6928 |  25-40    |

Which is indeed much better i.e. ~x2 times

**Comment by jkotan** [comment]
@ingvord if you like to speed-up your tests more you can use `dataset.resize({i + 1});` instead of `dataset.extent(0, 1);` The first one is faster, i.e. a raw method without additional checks,  but you have to know the target size of the dataset, i.e.
```c++
    for(size_t i = 0; i < 1000000; i++) {
      dataset.resize({i + 1});
      auto file_space = dataset.dataspace();
      selection.offset(0, i);
      file_space.selection(hdf5::dataspace::SelectionOperation::SET,selection);
      dataset.write(static_cast<long int>(std::rand()), dtype, simple_space, file_space, dtpl);
    }
```

**Comment by Ingvord** [comment]
@jkotan 

Thank you very much for your help!

I have implemented some of the tactics discussed here and in [h5cpp-510](https://github.com/ess-dmsc/h5cpp/issues/510) and we experience acceptable performance at our beamlines now.
